docs/src/writing-constructs.rst

```diff
@@ -0,0 +1,169 @@
+   either express or implied. See the License for the specific language governing permissions and
+   limitations under the License.
+
+.. _writing_constructs:
+
+#########################
+Writing |cdk| Constructs
+#########################
+
+This topic provides some tips writing idiomatic new constructs for the |cdk|.
+The tips here apply equally to constructs written for inclusion in the AWS
+Construct Library, purpose-built constructs to achieve a well-defined goal,
+or constructs that serve as building blocks for assembling your cloud
+applications.
+
+General Design Priciples
+========================
+
+* Favor composition over inheritance; most of the constructs should directly
+  extend the **Construct** class instead of some other construct. Inheritance
+  should mainly be used to allow polymorphism. Typically, you'll add a child
+  construct and expose any of its APIs and properties in the parent construct.
+* Provide defaults for everything that a reasonable guess can be made for;
+  ideally, props should be optional and **new MyAwesomeConstruct(this, "Foo")**
+  should be enough to set up a reasonable variant of the construct. This does
+  not mean that the user should not have the opportunity to customize! Rather,
+  it means that the specific parameter should be optional and set to a
+  reasonable value if not supplied. This may involve creating other resources
+  as part of constructing this construct. For example, all resources that
+  require a role allow passing in a **Role** object, but if the user does
+  not supply one an appropriate **Role** object is constructed in-place.
+* Use contextual defaulting between properties; the value of one property may
+  affect sensible defaults for other properties. *Example*: enableDnsHostnames
+  and and enableDnsSupport.  dnsHostnames requires dnsSupport, only yell if the
+  user has *explicitly* disabled dnsSupport, but tried to enable dnsHostnames.
+  A user expects things to just work - make that the case unless they tell us
+  otherwise.
+* Make the user thinkg about concepts, not implementation detail; for example,
+  if establishing an association between two resources (such as a **Topic**
+  and a **Queue**) requires multiple steps (in this case, creating a
+  **Subscription** but also setting appropriate IAM permissions), make
+  both things happen in a single call (to **subscribeQueue()**). The user
+  should be mentally thinking about the result they're trying to achieve,
+  not the various steps necessary to get there.
+* Make sure you are not introducing new concepts or terminology. For example
+  don't be temped to replace SQS's **dataKeyReusePeriod** with **keyRotation**
+  because it will be hard for people to diagnose problems. They won't be able to
+  just search for *sqs dataKeyReuse* and find topics on it.
+* Indicate units of measurement in property names that don't use a strong type.
+  Use **milli**, **sec**, **min**, **hr**, **Bytes**, **KiB**, **MiB**, **GiB**,
```

Comment:
Some examples (of where we do this) would be welcome.

Reply:
I don't really want to do that, as we only have seconds :)